### PR TITLE
NSX-V distributed firewall resource and data source

### PR DESCRIPTION
**Pull request by dataclouder**
This PR adds 
* a [resource for `vcd_nsxv_distributed_firewall`](https://github.com/dataclouder/terraform-provider-vcd/blob/distributed-firewall-1/website/docs/r/nsxv_distributed_firewall.html.markdown)
* a [data source for `vcd_nsxv_distributed_firewall`](https://github.com/dataclouder/terraform-provider-vcd/blob/distributed-firewall-1/website/docs/d/nsxv_distributed_firewall.html.markdown)
* a [data source for `vcd_nsxv_application_finder`](https://github.com/dataclouder/terraform-provider-vcd/blob/distributed-firewall-1/website/docs/d/nsxv_application_finder.html.markdown), to find applications and application groups to use with the NSX-V distributed firewall.
* a [data source for `vcd_nsxv_application`](https://github.com/dataclouder/terraform-provider-vcd/blob/distributed-firewall-1/website/docs/d/nsxv_application.html.markdown) to read an application to use with the NSXV distributed firewall
* a [data source for `vcd_nsxv_application_group`](https://github.com/dataclouder/terraform-provider-vcd/blob/distributed-firewall-1/website/docs/d/nsxv_application_group.html.markdown) to read an application group to use with the NSX-V distributed firewall
* a new property for `vcd_org_vdc` resource and data source: `enable_distributed_firewall`, which allows sys admins to enable the firewall when creating a VDC.

**NOTE**: As per product specs, no default rules are added. Users are free to set or remove rules without constraints. If users want to set a default rule (such as a "fallback deny all") they must add it explicitly.

### CODE REVIEW

**Review by lvirbalas — CHANGES_REQUESTED** [review]
Cool big PR! Mainly suggestions for polishing inline.

**Review by lvirbalas — APPROVED** [review]
Thanks for tune-ups!

File: vcd/resource_vcd_nsxv_distributed_firewall.go [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+		}
+		rulesList = append(rulesList, ruleMap)
+	}
+
+	return d.Set("rule", rulesList)
+}
+
+// stringOnNotNil returns the contents of a string pointer
+// if the pointer is nil, returns an empty string
+func stringOnNotNil(p *string) string {
```
Comment (adambarreiro) [inline vcd/resource_vcd_nsxv_distributed_firewall.go:599]: Just a suggestion, would it make sense to have this one together with `stringPtrOrNil` in structure.go? As it is the opposite function.

File: vcd/resource_vcd_nsxv_distributed_firewall.go [hunk 2]
```diff
@@ -0,0 +1,647 @@
...
+	if p == nil {
+		return ""
+	}
+	return *p
+}
+
+func resourceVcdNsxvDistributedFirewallImport(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
```
Comment (adambarreiro) [inline vcd/resource_vcd_nsxv_distributed_firewall.go:606]: I see in other resources that we usually document how the Import works (example:` resourceVcdOrgVdcImport`), would it make sense to do the same here?

File: website/docs/d/nsxv_application.html.markdown [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+layout: "vcd"
+page_title: "VMware Cloud Director: vcd_nsxv_application"
+sidebar_current: "docs-vcd-data-source-nsxv-application"
+description: |-
+  Provides a VMware Cloud Director data source for reading NSX-V distributed firewall applications
+---
+
+# vcd\_nsxv\_application
+
+Provides a VMware Cloud Director NSX-V distributed firewall application data source used to read an existing application.
```
Comment (adambarreiro) [inline website/docs/d/nsxv_application.html.markdown:11]: Maybe we could put line 6 contents here, looks clearer to me imho.

File: website/docs/d/nsxv_application_finder.html.markdown [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+page_title: "VMware Cloud Director: vcd_nsxv_application_finder"
+sidebar_current: "docs-vcd-data-source-nsxv-application-finder"
+description: |-
+  Provides a VMware Cloud Director data source for searching NSX-V distributed firewall applications and application groups
+---
+
+# vcd\_nsxv\_application_finder
+
+Provides a VMware Cloud Director NSX-V distributed firewall applications and application groups finder
+used to retrieve existing applications by regular expressions.
```
Comment (adambarreiro) [inline website/docs/d/nsxv_application_finder.html.markdown:12]:
```suggestion
used to retrieve existing applications and application groups by regular expressions.
```
or maybe better

```suggestion
used to retrieve existing ones by regular expressions.
```

**Review by adambarreiro — COMMENTED** [review]
First scan, just some nits and doubts.

**Review by Didainius — COMMENTED** [review]
Looks good while testing. Some inline comments.
I also saw some `TODO` items - are these fine to leave now, or should any of them be reevaluated?

File: vcd/datasource_vcd_org_vdc.go [hunk 5]
```diff
@@ -254,5 +261,14 @@ func datasourceVcdOrgVdcRead(_ context.Context, d *schema.ResourceData, meta int
 		return diag.FromErr(err)
 	}
 
+	dSet(d, "enable_distributed_firewall", false)
+	if adminVdc.IsNsxv() {
+		dfw := govcd.NewNsxvDistributedFirewall(&vcdClient.Client, adminVdc.AdminVdc.ID)
+		enabled, err := dfw.IsEnabled()
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/datasource_vcd_org_vdc.go:267]: We have had bad experience when trying to handle possible errors due to user privileges. I have also tested on an user with fewer privileges, and the error messages are clear enough for  users to invoke more privileges from their providers.

File: vcd/resource_vcd_nsxv_distributed_firewall.go [hunk 6]
```diff
@@ -0,0 +1,647 @@
...
+	err = dfwRulesToResource(dfw.Configuration.Layer3Sections.Section.Rule, d)
+	if err != nil {
+		return diag.Errorf("error setting distributed firewall state: %s", err)
+	}
+	util.Logger.Println("[NSX-V DFW END]")
+
+	return nil
+}
+
+func resourceVcdNsxvDistributedFirewallCreateUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}, origin string) diag.Diagnostics {
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/resource_vcd_nsxv_distributed_firewall.go:286]: Good catch. I wanted to use it to provide better error messages. Will fix it one way or other

File: vcd/resource_vcd_nsxv_distributed_firewall.go [hunk 7]
```diff
@@ -0,0 +1,647 @@
...
+func resourceVcdNsxvDistributedFirewallDelete(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	vdcId := d.Get("vdc_id").(string)
+	if vdcId == "" {
+		vdcId = d.Id()
+	}
+	dfw := govcd.NewNsxvDistributedFirewall(&vcdClient.Client, vdcId)
+	util.Logger.Printf("[INFO] disabling NSX-V distributed firewall for VDC %s\n", vdcId)
+	err := dfw.Disable()
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/resource_vcd_nsxv_distributed_firewall.go:330]: It removes everything, including layer 2 rules, which we don't handle in this resource, as per product specs

File: vcd/datasource_vcd_org_vdc.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Didainius) [reply vcd/datasource_vcd_org_vdc.go:267]: My main point here is that it is important to have the data source working even if one cannot read DFW status as it is the only tool to retrieve `vdc_id` and not everyone can have all rights needed. It didn't fail for me with "vApp user" role  so hopefully this is good

File: vcd/resource_vcd_nsxv_distributed_firewall.go [hunk 8]
```diff
@@ -0,0 +1,647 @@
...
+
+func resourceVcdNsxvDistributedFirewallCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	return resourceVcdNsxvDistributedFirewallCreateUpdate(ctx, d, meta, "create")
+}
+
+func resourceVcdNsxvDistributedFirewallUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	return resourceVcdNsxvDistributedFirewallCreateUpdate(ctx, d, meta, "update")
+}
+
+func resourceVcdNsxvDistributedFirewallDelete(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/resource_vcd_nsxv_distributed_firewall.go:321]: done

File: vcd/datasource_vcd_nsxv_application_finder.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+	for _, application := range applications {
+		item := map[string]string{
+			"name":  application.Name,
+			"type":  application.Type.TypeName,
+			"value": application.ObjectID,
+		}
+		result = append(result, item)
+	}
+
+	for _, sg := range applicationGroups {
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/datasource_vcd_nsxv_application_finder.go]: fixed

File: vcd/resource_vcd_nsxv_distributed_firewall.go (hunk 2 above)
Comment (dataclouder) [reply vcd/resource_vcd_nsxv_distributed_firewall.go:602]: done

File: vcd/datasource_vcd_nsxv_distributed_firewall.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+							Computed:    true,
+							Description: "List of destination traffic for this rule. Leaving it empty means 'any'",
+							Elem: &schema.Resource{
+								Schema: computedMap(destinationDef().Schema),
+							},
+						},
+						"exclude_destination": {
+							Type:        schema.TypeBool,
+							Computed:    true,
+							Description: "If set, reverses the content of the destination elements",
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/datasource_vcd_nsxv_distributed_firewall.go]: fixed

File: vcd/datasource_vcd_org_vdc.go [hunk 11]
```diff
@@ -224,6 +225,11 @@ func datasourceVcdOrgVdc() *schema.Resource {
 				Computed:    true,
 				Description: "ID of NSX-T Edge Cluster (provider vApp networking services and DHCP capability for Isolated networks)",
 			},
+			"enable_distributed_firewall": {
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/datasource_vcd_org_vdc.go]: changed

File: vcd/datasource_vcd_org_vdc.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/datasource_vcd_org_vdc.go:267]: noted

File: vcd/resource_vcd_nsxv_distributed_firewall.go (hunk 1 above)
Comment (dataclouder) [reply vcd/resource_vcd_nsxv_distributed_firewall.go]: yes. Moved

File: website/docs/d/nsxv_application.html.markdown (hunk 3 above)
Comment (dataclouder) [reply website/docs/d/nsxv_application.html.markdown]: done

File: website/docs/d/nsxv_application_finder.html.markdown (hunk 4 above)
Comment (dataclouder) [reply website/docs/d/nsxv_application_finder.html.markdown]: used the second suggestion

File: website/docs/d/nsxv_application_finder.html.markdown [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+  "vdc_id" = "urn:vcloud:vdc:e5680ceb-1c15-48a8-9a54-e0bbc6fe909f"
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `vdc_id` - (Required) The ID of VDC to use.
+* `search_expression` - (Required) The regular expression that will be used to search the applications
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply website/docs/d/nsxv_application_finder.html.markdown]: done

File: vcd/resource_vcd_nsxv_distributed_firewall.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (adambarreiro) [reply vcd/resource_vcd_nsxv_distributed_firewall.go:330]: Thanks for the explanation!

File: vcd/datasource_vcd_nsxv_distributed_firewall.go [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+			},
+			"enabled": {
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "When true, it enables the NSX-V distributed firewall",
+			},
+			"rule": {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "Ordered list of distributed firewall rules. Will be considered only if `enabled` is true",
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/datasource_vcd_nsxv_distributed_firewall.go]: No need for it. Removed

File: vcd/datasource_vcd_nsxv_distributed_firewall.go [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+		Schema: map[string]*schema.Schema{
+			"vdc_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The ID of VDC",
+			},
+			"enabled": {
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "When true, it enables the NSX-V distributed firewall",
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/datasource_vcd_nsxv_distributed_firewall.go]: Changed

File: vcd/datasource_vcd_nsxv_distributed_firewall.go [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+						},
+						"packet_type": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "Packet type of the rule (any, ipv4, ipv6)",
+						},
+						"source": {
+							Type:        schema.TypeSet,
+							Computed:    true,
+							Description: "List of source traffic for this rule. Leaving it empty means 'any'",
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/datasource_vcd_nsxv_distributed_firewall.go]: Changed

File: vcd/datasource_vcd_nsxv_distributed_firewall.go [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+						},
+						"exclude_source": {
+							Type:        schema.TypeBool,
+							Computed:    true,
+							Description: "If set, reverses the content of the source elements",
+						},
+						"destination": {
+							Type:        schema.TypeSet,
+							Computed:    true,
+							Description: "List of destination traffic for this rule. Leaving it empty means 'any'",
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/datasource_vcd_nsxv_distributed_firewall.go]: Changed

File: .changes/v3.9.0/988-improvements.md [hunk 17]
```diff
@@ -0,0 +1 @@
+* `vcd_org_vdc` includes a property `enable_distributed_firewall` to enable or disable a NSX-V distributed firewall [GH-988]
```
Comment (lvirbalas) [inline .changes/v3.9.0/988-improvements.md:1]: The name of the property has became outdated.

File: website/docs/d/nsxv_application.html.markdown [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+layout: "vcd"
+page_title: "VMware Cloud Director: vcd_nsxv_application"
+sidebar_current: "docs-vcd-data-source-nsxv-application"
+description: |-
+  Provides a VMware Cloud Director data source for reading NSX-V distributed firewall applications
+---
+
+# vcd\_nsxv\_application
+
+Provides a VMware Cloud Director data source for reading NSX-V distributed firewall applications
```
Comment (lvirbalas) [inline website/docs/d/nsxv_application.html.markdown:11]: Nit, it looks strange without a dot in this place.

```suggestion
Provides a VMware Cloud Director data source for reading NSX-V distributed firewall applications.
```

File: website/docs/d/nsxv_application_group.html.markdown [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+layout: "vcd"
+page_title: "VMware Cloud Director: vcd_nsxv_application_group"
+sidebar_current: "docs-vcd-data-source-nsxv-application-group"
+description: |-
+  Provides a VMware Cloud Director data source for reading NSX-V Distributed Firewall application groups
+---
+
+# vcd\_nsxv\_application\_group
+
+Provides a VMware Cloud Director NSX-V Distributed Firewall data source used to read an existing application group
```
Comment (lvirbalas) [inline website/docs/d/nsxv_application_group.html.markdown:11]: Nit, it looks strange without a dot in this place.

Comment (lvirbalas) [reply website/docs/d/nsxv_application_group.html.markdown:11]: Thanks for the change, let's also use dot in this place.

```suggestion
Provides a VMware Cloud Director NSX-V Distributed Firewall data source used to read an existing application group.
```

File: .changes/v3.9.0/988-improvements.md (hunk 17 above)
Comment (dataclouder) [reply .changes/v3.9.0/988-improvements.md]: done

File: website/docs/d/nsxv_application_group.html.markdown (hunk 19 above)
Comment (dataclouder) [reply website/docs/d/nsxv_application_group.html.markdown]: fixed

File: website/docs/d/nsxv_application.html.markdown (hunk 18 above)
Comment (dataclouder) [reply website/docs/d/nsxv_application.html.markdown]: fixed

File: website/docs/d/nsxv_application_group.html.markdown [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+layout: "vcd"
+page_title: "VMware Cloud Director: vcd_nsxv_application_group"
+sidebar_current: "docs-vcd-data-source-nsxv-application-group"
+description: |-
+  Provides a VMware Cloud Director data source for reading NSX-V distributed firewall application groups
+---
+
+# vcd\_nsxv\_application\_group
+
+Provides a VMware Cloud Director NSX-V distributed firewall application used to read an existing application group
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply website/docs/d/nsxv_application_group.html.markdown]: changed

File: website/docs/d/nsxv_application_finder.html.markdown [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+page_title: "VMware Cloud Director: vcd_nsxv_application_finder"
+sidebar_current: "docs-vcd-data-source-nsxv-application-finder"
+description: |-
+  Provides a VMware Cloud Director data source for searching NSX-V distributed firewall applications and application groups
+---
+
+# vcd\_nsxv\_application_finder
+
+Provides a VMware Cloud Director NSX-V distributed firewall applications and application groups finder
+used to retrieve existing applications by regular expressions
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply website/docs/d/nsxv_application_finder.html.markdown]: fixed

File: website/docs/d/nsxv_application.html.markdown [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+  "vdc_id" = "urn:vcloud:vdc:e5680ceb-1c15-48a8-9a54-e0bbc6fe909f"
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `vdc_id` - (Required) The ID of VDC to use.
+* `name` - (Required) The name of the application.
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply website/docs/d/nsxv_application.html.markdown]: fixed

File: website/docs/d/nsxv_application.html.markdown [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+layout: "vcd"
+page_title: "VMware Cloud Director: vcd_nsxv_application"
+sidebar_current: "docs-vcd-data-source-nsxv-application"
+description: |-
+  Provides a VMware Cloud Director data source for reading NSX-V distributed firewall applications
+---
+
+# vcd\_nsxv\_application
+
+Provides a VMware Cloud Director NSX-V distributed firewall application used to read an existing application
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply website/docs/d/nsxv_application.html.markdown]: fixed

**Review by Didainius — APPROVED** [review]

File: vcd/datasource_not_found_test.go [hunk 24]
```diff
@@ -79,7 +79,8 @@ func testSpecificDataSourceNotFound(dataSourceName string, vcdClient *VCDClient)
 				t.Skip(`Works only with system admin privileges`)
 			}
 		// vcd_resource_list and vcd_resource_schema don't search for real entities
```
Comment (adezxc) [inline vcd/datasource_not_found_test.go:81]: nit: update comment

**Review by adambarreiro — APPROVED** [review]
Amazing work, thanks!

**Review by adezxc — APPROVED** [review]

File: vcd/datasource_not_found_test.go (hunk 24 above)
Comment (dataclouder) [reply vcd/datasource_not_found_test.go:81]: Thanks. Updated

**Pull request merged**